api/v1alpha1/openstackserver_types.go

```diff
@@ -0,0 +1,186 @@
+	// +listType=map
+	// +listMapKey=key
+	// +optional
+	ServerMetadata []infrav1.ServerMetadata `json:"serverMetadata,omitempty"`
+
+	// Tags which will be added to the machine and all dependent resources
+	// which support them. These are in addition to Tags defined on the
+	// cluster.
+	// Requires Nova api 2.52 minimum!
+	// +listType=set
+	Tags []string `json:"tags,omitempty"`
+
+	// Trunk is a flag to indicate if the server instance is created on a trunk port or not.
+	// +optional
+	Trunk optional.Bool `json:"trunk,omitempty"`
+
+	// UserData is the user data to be injected into the server instance.
+	// +optional
+	UserData optional.String `json:"userData"`
+}
+
+// OpenStackServerStatus defines the observed state of OpenStackServer.
+type OpenStackServerStatus struct {
+	// Ready is true when the OpenStack server is ready.
+	// +kubebuilder:default=false
+	Ready bool `json:"ready"`
+
+	// InstanceID is the ID of the server instance.
+	// +optional
+	InstanceID optional.String `json:"instanceID,omitempty"`
+
+	// InstanceState is the state of the server instance.
+	// +optional
+	InstanceState *infrav1.InstanceState `json:"instanceState,omitempty"`
+
+	// Addresses is the list of addresses of the server instance.
+	// +optional
+	Addresses []corev1.NodeAddress `json:"addresses,omitempty"`
+
+	// Resolved contains parts of the machine spec with all external
+	// references fully resolved.
+	// +optional
+	Resolved *ResolvedServerSpec `json:"resolved,omitempty"`
+
+	// Resources contains references to OpenStack resources created for the machine.
+	// +optional
+	Resources *ServerResources `json:"resources,omitempty"`
+
+	// Conditions defines current service state of the OpenStackServer.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+// +genclient
+// +genclient:Namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +kubebuilder:resource:path=openstackservers,scope=Namespaced,categories=cluster-api,shortName=oss
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="InstanceState",type="string",JSONPath=".status.instanceState",description="OpenStack instance state"
+// +kubebuilder:printcolumn:name="Ready",type="string",JSONPath=".status.ready",description="OpenStack instance ready status"
+// +kubebuilder:printcolumn:name="InstanceID",type="string",JSONPath=".status.instanceID",description="OpenStack instance ID"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description="Time duration since creation of OpenStack instance"
+
+// OpenStackServer is the Schema for the openstackservers API.
+type OpenStackServer struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   OpenStackServerSpec   `json:"spec,omitempty"`
+	Status OpenStackServerStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// OpenStackServerList contains a list of OpenStackServer.
```

Comment:
Note for the reviewers: I didn't add ` OpenStackServerTemplate` because `OpenStackServer` isn't supposed to be user-exposed, even though an user could create that resource and the server would be managed by CAPO.
However we should recommend our users to use something like https://github.com/k-orc/openstack-resource-controller instead.